### ISSUE TITLE
Gracefully recover when utxo already spent

### ISSUE THREAD

**Issue by chris-belcher** [JoinMarket-Org/joinmarket#74]
I came across a situation just now which caused a yield generator to crash.

The yieldgen Y had a deal done with a taker A, after Y had sent the signature but before A broadcast the tx, a second taker B started a coinjoin deal with Y.

Meanwhile A had broadcast the tx and the bitcoin core updated it's chain state

Then when Y went to do a sanity check, it saw the utxo was already spent, and paniced and quit.

The sanity checking code is here https://github.com/chris-belcher/joinmarket/blob/master/lib/maker.py#L44

What the bot should actually do is gracefully remove the spent utxo from its internal wallet and call oid_to_order() again. If not possible to fill the order it should send back an error and take the reputation hit as in https://github.com/chris-belcher/joinmarket/issues/57 . This means sending back an error in the early stages of the deal should be punished less harshly than timing out with no messages in the later stages.

Note: the other sanity checking code that checks the utxos are labelled with the correct value, should remain.

**Comment by chris-belcher** [comment]
this crash was probably caused by this issue https://github.com/chris-belcher/joinmarket/issues/69 if @domob1812 still had the log file we could make sure.

**Comment by chris-belcher** [comment]
Someone is apparently exploiting this in the wild. Time to actually fix this.

**Comment by AdamISZ** [comment]
Is anyone working on this? I can take a look.

**Comment by AdamISZ** [comment]
Analyzing it; it seems that the time window for this to occur is tiny. If unconfirmfun has run before the start of CoinJoinOrder.__init__ , so that the order's been cancelled, the taker will just receive the 'oid not found' error. So the walletnotify and unconfirmfun execution has to complete between lines 25 and 44. Have I understood that right?

**Comment by cqv2** [comment]
There are two ways this can cause problems:
-1 pushtx is having problems and the taker is submitting the transaction manually again with delay.
-2 The taker want to shoot down a maker. He intentionally waits before submitting the transaction.
In both cases, if the taker waits for longer than 'unconfirm_timeout' the transaction is ignored by the maker and the script will try to use the spent output again with the next taker.

**Comment by AdamISZ** [comment]
@cqv2 but what I'm saying in my above comment is, in order to create the *specific* result as described in @chris-belcher 's OP, you need the transaction arriving in the wallet (and thence triggering the unconfirm_fun, which modifies the orders) between the call to maker.orderlist (line 25) and the call to query_utxo_set. If it happens before reading the orderlist, the order will have been cancelled and you'll get the error 'oid not found', and if it happens after query_utxo_set is called, the new taker will just use the old (stale) utxo list and try to create a transaction, which results in pushtx failure. Neither of these is the 'panic and quit' scenario described, which happens on line 44/45 if utxos have been chosen that no longer are unspent.

**Comment by AdamISZ** [comment]
Analyzing a bit more, it seems that this particular scenario isn't actually possible, at least on the current codebase.

The CoinJoinOrder constructor is wrapped in a wallet_unspent_lock acquisition. (https://github.com/chris-belcher/joinmarket/blob/master/lib/maker.py#L194-L198). 

I tested by using file semaphores to enforce that the unconfirm_callback is only called once the code reaches [this](https://github.com/chris-belcher/joinmarket/blob/master/lib/maker.py#L28) line. In other words, I have the unconfirm_callback in the Notifier thread wait for that line, then allow it to execute and force the execution of the main thread to wait until unconfirm_callback completes. 

But when unconfirm_callback is called, it tries to acquire the lock here: https://github.com/chris-belcher/joinmarket/blob/master/lib/maker.py#L103 , and fails/hangs/deadlocks.

That's my reading of it for now (that the precise error condition explained in the first post here can't happen because, without using semaphores, the unconfirm_callback would wait until the CoinJoinOrder constructor had completed before accessing the wallet unspent). @chris-belcher let me know if I've missed something.

**Comment by chris-belcher** [comment]
> Analyzing it; it seems that the time window for this to occur is tiny. If unconfirmfun has run before the start of CoinJoinOrder.init , so that the order's been cancelled, the taker will just receive the 'oid not found' error. So the walletnotify and unconfirmfun execution has to complete between lines 25 and 44. Have I understood that right?

The order doesn't always get cancelled. It will get cancelled if maker only has one UTXO which is now locked up in the transaction. For a maker with more UTXOs the orders will remain, maybe being modified with a new size but still wont return 'oid not found' if filled.

**Comment by AdamISZ** [comment]
@chris-belcher thanks. that's what i was missing. i have to rejig my tests to make that happen.

**Comment by AdamISZ** [comment]
So the unconfirm_callback will remove the utxos from txs that appear on the network irrespective of whether it then !cancels any orders, right? So if taker B appears and does his !fill after the tx was seen on the network, then the utxos from the first tx are already removed from the wallet, no? Then self.utxos as returned from oid_to_order would not contain any such utxos, right?

I have done some crude tests and get behaviour that seems to match this, i.e. after 1st tx seen on network, its utxos are removed, and a new order fill occurring here (before first tx is confirmed in a block) runs OK (in as much as it doesn't try to reuse no longer available utxos).

**Comment by chris-belcher** [comment]
Yes that's right. The problem comes up if the taker waits a while after it got !sig but before it does pushtx(). Maybe because another maker is really slow, maybe an error happened with pushing and a human did pushtx() manually, or maybe taker is malicious and wants to crash the maker.

**Comment by AdamISZ** [comment]
If we have this sequence:
A !fills
Y passes sig
A waits for some time (doesn't broadcast), meanwhile:
   B !fills
   Y passes sig (hasn't got a notify thread call from A so no problem)
   B pushes
Later: A pushes and broadcasts, but it fails.

this doesn't trigger the described error. I've just done this test.

But if 
A !fills 
Y passes sig
A waits for some time (doesn't broadcast), meanwhile:
 B !fills
 NOW A broadcasts and notifies, unconfirm_fun calls. But it has to wait on the wallet_unspent lock. So it doesn't trigger the error? (this is what I was describing a few messages up the thread)

**Comment by chris-belcher** [comment]
That can't be right. That would mean wallet_unspent_lock is acquired after !sig is sent and only released when unconfirm_fun is called, which might be never. Locks are not meant to be held for possibly hours.
Reading the source.

**Comment by chris-belcher** [comment]
It seems you're right. Hold on.

**Comment by chris-belcher** [comment]
OK. I think this.
The wallet_unspent_lock only controls access to the wallet.unspent list in python, not to the Bitcoin Core chainstate or blockr.io chainstate which is being queried with bc_interface.query_utxo_set()

A !fills 
Y passes sig
A waits for some time (doesn't broadcast)
B !fills
Y acquires wallet_unspent_lock, creates CoinJoinOrder and starts to execute __init__()
A broadcasts tx
Y's bitcoin core sees A's transaction and updates its chainstate, and also calls -walletnotify
Y's unconfirm_fun() is called but hangs when trying to acquire wallet_unspent_lock, so it doesn't yet call wallet.remove_old_utxos()
Y's CoinJoinOrder calls oid_to_order() which returns the same UTXOs as used in A's tx
Y's CoinJoinOrder checks the UTXOs by calling query_utxo_set(), finds them already spent and calls exit()

I think that should cause a crash. You're right that this a very small window for this to happen. I definitely remember a bug happening to my yieldgenerator which caused me to open this issue, and I'd be amazed if it really was this.
When I get home I'll search through my debug logs, the one with the crash should still be there.

**Comment by AdamISZ** [comment]
Yes, what you describe is my understanding (hence it's not a lock held for hours, it's only spawned by the tx broadcast, and it'll only be a lock conflict if the two things happen at the same time, so yeah seems to be an ultra unlikely occurrence).

So it seems like you need this sequence of events: B's fill order is called. Then the walletnotify happens a tiny time after, so it hangs at the unspent lock, and so as you say, the self.utxos will be out of sync with the Core wallet.

If on the other hand the sequence is the opposite (walletnotify grabs the lock first) then we don't get the crash.

I think my previous test with semaphores failed because I did it the wrong way round. Will try again.

**Comment by AdamISZ** [comment]
Running a few more tests (as alluded to above), I found this: 

If B calls query_utxo_set and thus rpc 'gettxout' on the utxo that was spent in A's tx, and which A has broadcast and is in the mempool, but not yet confirmed, that utxo is still returned, even though the transaction spending it is in the mempool. Thus B sees no problem and pushes the double spending tx, which then fails.

This is what I see from putting debug statements in the query_utxo_function (and using the semaphore method mentioned to trigger the scenario). Could be wrong though.

**Comment by AdamISZ** [comment]
So, investigating further:

If the third argument to gettxout is set to 'false', as it currently is in the code, txs in the mempool are not included. That's why query_utxo_set is not paying attention to the broadcast but not yet committed tx from A.

I reset that parameter to 'true' and in this way I was able to reproduce the crash.

So according to my current understanding, this crash can't occur in the current version of the code.

**Comment by chris-belcher** [comment]
Ok. Seems you're right.

**Comment by CohibAA** [comment]
[2015/08/08 01:51:27] wrongly using an already spent utxo. utxo_data = [None]

This was using the new-yield-gen-algo bot, but no other changes to master.  I'll try to get the logs posted (would prefer to PM relevant pieces to @chris-belcher  and/or @AdamISZ if possible), but it seems this is still possible/happening.  From a quick review of the log, it looks like the same taker initiated !fill for two different amounts, about 15 minutes apart.  The listening thread was active, but had not received a confirmation on the first fill, if I read it correctly.

**Comment by chris-belcher** [comment]
For anything like this you can use my email address which is belcher@riseup.net and pgp key fingerprint is here https://twitter.com/chris_belcher_/status/559879313622061056

**Comment by AdamISZ** [comment]
I have found a scenario where I can reproducibly generate this error. It is artificial though; but still it might help give insight. It works like this:

In the regtest script, I start a yield generator bot, and then a sendpayment bot. They both choose appropriate utxos for the join, and do the transaction. The sendpayment bot then shuts down as usual. Then another sendpayment bot starts, attempts a join, and the yieldgen erroneously chooses the *same* utxo as before. When it calls the 'gettxout' on that utxo, 'None' is returned and we get this "wrongly using an already spent utxo."

The reason: in the current version of the code there is an allowance made for more than one bot receiving notifications from the same bitcoind instance [here](https://github.com/chris-belcher/joinmarket/blob/master/lib/blockchaininterface.py#L326-L327). It relies on the fact that the code will bind to port numbers incrementally when other ports are unavailable (i.e. 62602, 62603 etc..). Thus, that chunk of code can ensure that all bots get the notification. The problem here is that if the sendpayment bot gets the notification first, because it has port 62602, but it shuts down immediately on pushing the transaction, the yieldgen bot will never get the notification, and will have no reason to believe that the the transaction has been seen on the network, and so will not remove that utxo from its wallet. Hence the above crash.

I have a solution for this which I needed to do tumbler tests (much more demanding, 7 bots) - I pass a shell script in the bitcoin.conf which calls the curl to all the relevant ports simultaneously to avoid any dependencies.

Obviously this wallet notify problem doesn't apply to normal joinmarket. I'll have to review what was said before, but it seems like I got over-interested in the matter of the wallet_unspent lock and forgot to consider what happens if it never gets acquired anywhere, because the transaction is not even seen... If so, back to the start, and it just needs to be fixed as @chris-belcher originally said :)

**Comment by CohibAA** [comment]
@AdamISZ PM sent at bct.

The scenario you describe is very similar to what I interpret from the transaction flow in the log.  The takers, in both cases, attempted a second join with my yield-gen bot, which caused the bot to try to use the same utxo.  This is using the new-yield-gen-algo bot, to confirm.

Perhaps I have something mis-configured that is preventing the bot from updating its utxos or the wallet history?  What should I double-check if so?  Not sure if related, but this just reminded me... I've noticed I don't get details in the yigen-statement.csv, just lots of lines with commas.  Never thought much about it, tbh.

    2015/08/12 xx:xx:xx,,,,,,,Connected



Other than that, I was not the taker from my own bot, and nothing else is sharing the bitcoind connection, afaik.

**Comment by AdamISZ** [comment]
@CohibAA just in case you haven't seen, I've responded on bct. I can't see the logs.

**Comment by CohibAA** [comment]
Sorry, should be open now.

**Comment by AdamISZ** [comment]
@CohibAA 

Thanks. Addressing the second log file first:
there is a crash in the json decoder arising from line 358 : 

    txdata = json.loads(jsonstr)

The rest of the behaviour is understandable; since that notification thread crashes, it doesn't get a chance to update the wallet. So the wallet goes out of sync.

Since in the current new_yieldgen branch that line occurs on line 302, can you point me to the exact version of the code you're using?

I'll take a more careful look at the first log file, which appears to be a different case (no json decode crash).

**Comment by AdamISZ** [comment]
I think I'm right in saying you're using #141 ?

That might actually explain the json decode crash, if it's not .strip()ing the result properly (this was mentioned elsewhere, I'll look it up).
It's also conceivable that there's a problem with the calls to gettxout ("false" vs False?), or something similar, from that PR that might be causing the error in the first log file. But that's just a vague notion at the moment.

**Comment by CohibAA** [comment]
Right, this was with #141.  I (somehow) missed https://github.com/chris-belcher/joinmarket/pull/141#issuecomment-125048603 and having made those and another change or two, I haven't seen the utxo already spent crash again in the same scenario.  I'll move any further discussion to #141, thank you.

**Comment by AdamISZ** [comment]
OK. Once #141 is fully resolved and/or merged, we can come back and take another look at this. It probably makes sense to code the "recovery" that @chris-belcher originally described, but one is always reluctant to do that if one doesn't have a clear error path in mind; afaict there is no such error path in current master.

**Comment by chris-belcher** [comment]
My yield generator triggered this and ran sys.exit(0) last night, so I've got a log which shows what happened (not yet had time to look)

**Comment by CohibAA** [comment]
@chris-belcher If you were using the new-yieldgen-algo, keep in mind #195 also throws the utxo already spent crash under some conditions.  Either way, I look forward to your analysis.

**Comment by chris-belcher** [comment]
Unfortunately my crashes were because I forgot to run bitcoind with -walletnotify.

**Comment by riclas** [comment]
I believe I ran into some kind of problem like this as well. no concurrent orders though, i believe.
crash dump log:
http://pastebin.com/PE0ksR9m

**Comment by riclas** [comment]
this happened again today.... would you like me to open a new issue?

http://pastebin.com/thUhsB9j

**Comment by riclas** [comment]
and again... http://pastebin.com/eBtqsDXj

this never happened with blockr.

**Comment by riclas** [comment]
ok i changed back to blockr and it just happened... i guess i just wasn't having enough requests for it to happen.
